### ISSUE TITLE
No build steps visible in imports()

### ISSUE THREAD

**Issue by vadixidav** [conan-io/conan#2755]
Conan version `1.2.3`

In Conan `1.2` a feature was added to allow a developer to limit which stages of a build are executed. It has flags like `-c` and `-b` to control this. In the `build()` method you have access to `self.should_configure` and `self.should_build`. However, although the `imports()` function is heavily tied to the `build()` method in that it controls which artifacts are available at build and also cleans them up at the end, it does not have access to these members because they are only set immediately before `build()`.

I propose that the build step configuration variables also be made available in the `imports()` method since `imports()` is always called before build as part of the `conan build` development cycle step.

**Comment by memsharded** [comment]
Interesting. Could you please ellaborate a bit more on the use case, how would you use those variables inside the ``imports()`` method? Because in this case, the imports method is not executed just before the ``build()`` method. This variables only apply for the local flow with ``conan build`` command. If anything, the imports would be copied when ``conan install`` is executed, but at that moment you don't know if the build will be just a configure, a build... all the imports are needed to be copied.

**Comment by vadixidav** [comment]
I see. I thought `imports()` was invoked as part of `conan build`, but it is actually `conan install`.

The problem arises because the imports are not needed if the whole build process completes, but if a developer opens visual studio to edit, build, and run the project, certain resources need to exist from the current project and some imported resources from several dependencies. When the developer executes `conan install` and `conan build` the exports are not copied into the build folder, but they are when the package is created using, for instance, `conan create`. This makes it so a resource conflict exists where when the package is created with `conan create` because it will try to copy in the exported resource and the imported resource into the same location. In addition, I also require some exports to be present in the build folder, which creates further issues.

Currently, the only way for me to know if I am in `conan create` mode versus `conan build` mode in the `imports()` method is to use `hasattr(self, "build_folder")` inside the `imports()` method. This allows me to conditionally include the imports that are necessary when a developer is in the middle of working without also destroying the package exports when `conan create` is called. Then for some resources which are needed from the exports in the build folder, to avoid recursive directory copy issues I was including only if the developer uses `-c` or `-b` to stop the build process before the end so that I know.

I just thought of a temporary workaround while writing this. I now merge the "res" directory (where all these conflicts are happening) from the package source folder with the "res" directory which has the combined `exports` and `imports()` using [this Stack Overflow answer] which specifically avoids copying over files which have the same timestamp so I can avoid errors thrown due to the file being open by Conan due to the export (which seems to be there to protect the export?). This seems to be working, but all of these weird workarounds are obviously not what I want to be doing.

In conclusion, I need to know whether or not I am in `conan create`/`conan install <full package ref>` mode or not due to the difference in operation between builds which don't have exports and builds that do. I also need to know if a developer is executing `conan build` to perform a compile/debug cycle versus performing an intermediary build with the intention to create a package (in which case the resources aren't needed in the build folder). The latter thing I was planning to do by only importing if the developer specifies `-c` or `-b`, since that would imply they are planning to start a compile/debug cycle. So this ticket is specifically for the second problem, but in general I am having problems with all of these import problems at the same time.

**Comment by vadixidav** [comment]
Another solution would be to just have a flag to check if I am in a user build. It could be something like `self.user_build`.

**Comment by memsharded** [comment]
There are a couple of flags that might be useful:
- ``in_local_cache``: http://docs.conan.io/en/latest/reference/conanfile/attributes.html#in-local-cache
- ``develop``:  http://docs.conan.io/en/latest/reference/conanfile/attributes.html#develop

I think they allow to differentiate what you want.

> When the developer executes conan install and conan build the exports are not copied into the build folder, but they are when the package is created using, for instance, conan create. 

There exists the ``no_copy_source`` (http://docs.conan.io/en/latest/reference/conanfile/attributes.html#no-copy-source) flag that will instruct conan to not copy things to the build folder, even in the local cache, and things should keep working (of course if the build doesn't modify or pollute the original source folders)

Also, I am a bit confused about the imports and exports colliding. They are different things, they are not the same files, aren't they? If you could please ellaborate a bit further about this issue. Thanks!

**Comment by memsharded** [comment]
No problem, glad to help :)
